RecoEgamma/EgammaTools/src/ElectronEnergyCalibrator.cc

```diff
@@ -183,10 +183,18 @@ std::pair<float,float> ElectronEnergyCalibrator::calCombinedMom(reco::GsfElectro
 { 
   const float oldEcalEnergy = ele.ecalEnergy();
   const float oldEcalEnergyErr = ele.ecalEnergyError();
+
+  const auto oldP4 = ele.p4();
+  const float oldP4Err = ele.p4Error(reco::GsfElectron::P4_COMBINATION);
+  const float oldTrkMomErr = ele.trackMomentumError();
+ 
   setEcalEnergy(ele,scale,smear);
   const auto& combinedMomentum = epCombinationTool_->combine(ele);
+  
   ele.setCorrectedEcalEnergy(oldEcalEnergy);
   ele.setCorrectedEcalEnergyError(oldEcalEnergyErr);
+  ele.correctMomentum(oldP4,oldTrkMomErr,oldP4Err);
```

Comment:
is this update actually doing anything?

IIUC, the intent is to more appropriately reset back to the input ele content.
The jenkins comparisons in the reminiAOD of the 2017 show no differences.
I would expect changes there if this ```ele.correctMomentum(oldP4,oldTrkMomErr,oldP4Err)``` was not an effective no-op.
Is there a miss in monitoring or does this actually not lead to any differences in saved variables?

Reply:
It is actually doing something because of the way setCorrectedEcalEnergy works, it messes with the momentum. I found it when I was testing another modifier (was checking the modifier didnt change the energy but then it did). 

However practically speaking it doesnt do anything as its overriden with the corrected value later on : https://github.com/Sam-Harper/cmssw/blob/912df1243c293950c92e09b896385d2c2e67d990/RecoEgamma/EgammaTools/src/ElectronEnergyCalibrator.cc#L160 

And regardless, this version of the electron is discarded as we decided not to modify the electron. So you could set the momentum to zero and it would be okay.  So jenkins and the monitoring is fine.

So why did I change it if it has double no effect? Simply because the function had a side effect  (it subtly changed the momentum in an uncontrolled way) which even if it was harmless, could cause issues down the road. 

I fully admit, the entire way this works is a bit awful.

Reply:
given the repeated "awful" pattern of setting and resetting a reference electron, the more straightforward way would be to add a new combine method to the EpCombinationTool which can take the momentum and energy inputs with their uncertainties separate from the other parts of the electron.

BTW, the EgammaAnalysis/ElectronTools/interface/EpCombinationTool.h and RecoEgamma/EgammaTools/interface/EpCombinationTool.h define the same symbol
This situation is one unlucky move away from mysterious problems at run time.